### PR TITLE
[DataPipe] Adding HashJoinerIterDataPipe

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66204

Work in progress. Still need to add tests.

cc @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D31445594](https://our.internmc.facebook.com/intern/diff/D31445594)

### CODE REVIEW

**Review by ejguan — COMMENTED** [review]
The DataPipe looks reasonable to me. But, why not to implement it inside torchdata?

File: torch/utils/data/datapipes/iter/grouping.py [hunk 1]
```diff
@@ -319,3 +319,46 @@ def __iter__(self):
...
+    """
+
+    def __init__(self,
+                 source_iterdatapipe: IterDataPipe,
+                 map_datapipe: MapDataPipe,
+                 key_fn: Callable,
+                 merge_fn: Callable = tuple):
+        self.source_iterdatapipe = source_iterdatapipe
+        self.map_datapipe = map_datapipe
+        self.map_len = len(map_datapipe)
```
[the remark this answers is not in the document]
Comment (NivekT) [reply torch/utils/data/datapipes/iter/grouping.py:347]: We don't. I was trying something else but I will remove it now.

File: torch/utils/data/datapipes/iter/grouping.py [hunk 2]
```diff
@@ -319,3 +319,46 @@ def __iter__(self):
...
+        self.length = -1
+
+    def __iter__(self):
+        for item in self.source_iterdatapipe:
+            key = self.key_fn(item)
+            try:
+                map_item = self.map_datapipe[key]
+                yield self.merge_fn(item, map_item)
+            except (KeyError, IndexError):
+                raise KeyError(f"key_fn maps {item} to {key}, which is not a valid key in the given MapDataPipe.")
```
[the remark this answers is not in the document]
Comment (NivekT) [reply torch/utils/data/datapipes/iter/grouping.py:359]: We can, but I added it because I think a custom error message here will make it easier for users to understand what went wrong.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31445594).

**Comment by ejguan** [comment]
Just in case you missed my comment above, I think this PR should go into Data Repo.

> Note that because many features of the original DataLoader have been modularized into DataPipes, some now live as standard DataPipes in pytorch/pytorch rather than torchdata to preserve BC functional parity within torch.

**Comment by NivekT** [comment]
> > Note that because many features of the original DataLoader have been modularized into DataPipes, some now live as standard DataPipes in pytorch/pytorch rather than torchdata to preserve BC functional parity within torch.

@ejguan 

Does this imply there will not be any additional DataPipe added to Core from now on? Or is it simply stating that some DataPipes will persist in Core to preserve BC? I am thinking it is the latter.

In this particular case, I briefly talked to @vitaly-fedyunin yesterday (correct me if I'm wrong) and he mentioned that this one should go into Core. Personally, I think this DataPipe looks useful and it doesn't seem domain-specific. I don't have a strong preference tp where to put this.

I am open to add this to TorchData for now because the repo is still in a prototype stage, and then move it to Core later if it seems like a better fit.

**Comment by ejguan** [comment]
> Does this imply there will not be any additional DataPipe added to Core from now on? Or is it simply stating that some DataPipes will persist in Core to preserve BC? I am thinking it is the latter.

Emmm. I thought the DataPipes in Core are all related to functionality of DataLoader like shuffle, sharding, etc.

Anyways, I will let Vitaly to decide if we want to land it into Core since he has more context around this topic.

**Comment by NivekT** [comment]
> > Does this imply there will not be any additional DataPipe added to Core from now on? Or is it simply stating that some DataPipes will persist in Core to preserve BC? I am thinking it is the latter.
> 
> Emmm. I thought the DataPipes in Core are all related to functionality of DataLoader like shuffle, sharding, etc.
> 
> Anyways, I will let Vitaly to decide if we want to land it into Core since he has more context around this topic.

Agree - we can revisit this later. I will go ahead and add this to TorchData with a note that we will need to confirm the decision.
